### PR TITLE
feat: add nested transaction

### LINKED ISSUE #1505: Nesting transactions
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

I have a transactional method and from it i call other transactional methods. When i do this the other method create a new transaction, there are any option to create a nested transaction ? Something like the propagation options of [jpa/spring transaction](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/TransactionDefinition.html#PROPAGATION_MANDATORY).

The code above illustrate the question.

**Steps to reproduce or a small repository showing the problem:**

```ts 

@Transaction()
public async save(@RequestBody() band: Band, @TransactionRepository(Band) bandRepository: Repository<Band>) {
        /* new transaction 1 started */ 
        await bandRepository.save(band);
        await this.snooze(30000);
        await  this.saveMoreAwesomeBand(band.name, [band, band, band, band, band, band]);
        /* in this case the transaction 2 was commited but the 1 will be rollbacked */ 
        throw new Error('teste');
       //return await bandRepository.findAll();
    }

@Transaction()
private async saveMoreAwesomeBand(name: string, bands: Array<Band>, @TransactionRepository(Band) bandRepository?: Repository<Band>) {
        /* new transaction 2 started */
        for (let i = 0; i < bands.length; i++) {
            const band = bands[i];
            delete band.id;
            band.name = name + ' ' + i;
            await bandRepository.save(band);
        }
        /* transaction 2 commited */
    }


 ```

### PR DESCRIPTION
**Pull request by mortezaPRK**
### Description of change

Add nested transaction support.

Implemented following drivers:
* mysql
* postgres
* mariadb
* sqlite
* react-native
* nativescript
* sqljs
* oracle
* better-sqlite3
* capacitor

Following drivers also support nested transaction. but I'm not familiar with them:
* cockroachdb
* mssql
* aurora-data-api
* aurora-data-api-pg
* expo

May be someone else can implement it in another PR

Fixes: #1505 

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #1505`
- [x] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

> ~~This is WIP. if implementation is acceptable, I'll add other drivers and tests~~

### CODE REVIEW

File: src/driver/postgres/PostgresQueryRunner.ts [hunk 1]
```diff
@@ -31,6 +30,8 @@ import {MetadataTableType} from "../types/MetadataTableType";
  * Runs queries on a single postgres database connection.
  */
 export class PostgresQueryRunner extends BaseQueryRunner implements QueryRunner {
+    supportNestedTransactions = true;
+    private transaction_counter = 0;
```
Comment (Ginden) [inline src/driver/postgres/PostgresQueryRunner.ts:34]: Should be named `transactionDepth` probably.

**Comment by pleerock** [comment]
Thank you for your contribution, looks really good. Considering simplicity of the suggested solution, I just wanted to ask you if you have a time to investigate other drivers as well, because:

cockroachdb - should be same as postgres if they support it
aurora-data-api - should be same as mysql 
aurora-data-api-pg - should be same as postgres
expo -  should be same as sqlite
mssql - would be really nice to have since others are implemented

**Comment by mortezaPRK** [comment]
sure. I'll give it a try tomorrow

**Comment by mortezaPRK** [comment]
@pleerock 

I added other drivers except for mssql. I wasn't able to implement it :(

**Comment by AlexMesser** [comment]
thank you for contribution!

**Comment by EgzonIseini** [comment]
Hey thanks for the contribution @mortezaPRK , looks like exactly what I need. I was curious if the documentation has been updated by any chance on how to use this?

**Comment by xfishernet** [comment]
I try this example and it still doesn't work for me :-( 

https://github.com/typeorm/typeorm/issues/1505

**Comment by Gideon-Felt** [comment]
Is mssql still not capable of using this feature?

**Pull request merged**